docs/fundamentals/environments.rst

```diff
@@ -1,8 +1,129 @@
+
+When an ASP.NET 5 application starts, the ``Startup`` class is used to bootstrap the application, load its configuration settings, etc. (:doc:`learn more about ASP.NET startup <startup>`). However, if a class exists named ``Startup{EnvironmentName}``, e.g. ``StartupDevelopment``, and the ``ASPNET_ENV`` environment variable matches that name, then that ``Startup`` class is used instead. Thus, you could configure ``Startup`` for development, but have a separate ``StartupProduction`` that would be used when the app is run in production. Or vice versa.
+
+The following ``StartupDevelopment`` file from this articles sample project is run when the application is set to run in a Development environment:
+
+.. literalinclude:: environments/sample/src/Environments/StartupDevelopment.cs
+	:language: c#
+	:caption: StartupDevelopment.cs
+	:linenos:
+	:emphasize-lines: 5
+
+Run the application in development, and a welcome screen is displayed. The sample also includes a ``StartupStaging`` class:
+
+.. literalinclude:: environments/sample/src/Environments/StartupStaging.cs
+	:language: c#
+	:caption: StartupStaging.cs
+	:linenos:
+	:emphasize-lines: 6
+
+When the application is run with ``ASPNET_ENV`` set to ``Staging``, this ``Startup`` class is used, and the application will simply display a string stating it's running in a staging environment. The application's default ``Startup`` class will only run when the environment is not set to either ``Development`` or ``Staging`` (presumably, this would be when it is set to ``Production``, but you're not limited to only these three options. Also note that if no environment is set, the default ``Startup`` will run).
+
+In addition to using an entirely separate ``Startup`` class based on the current environment, you can also make adjustments to how the application is configured within a ``Startup`` class. The ``Configure()`` and ``ConfigureServices()`` methods support environment-specific versions similar to the ``Startup`` class itself, of the form ``Configure[Environment]()`` and ``Configure[Environment]Services()``. If you define a method ``ConfigureDevelopment()`` it will be called instead of ``Configure()`` when the environment is set to development. Likewise, ``ConfigureDevelopmentServices()`` would be called instead of ``ConfigureServices()`` in the same environment.
+
+Another approach you can use is to programmatically inspect the environment and use it to customize behavior within the ``Startup`` class. The ASP.NET 5 web site template in Visual Studio uses this approach to load environment-specific configuration files (if present) and to customize the app's error handling settings. In both cases, this behavior is achieved by referring to the currently specified environment by calling ``EnvironmentName`` or ``IsEnvironment`` on an instance of ``IHostingEnvironment`` passed into the appropriate method.
```

Comment:
I'd like to move this content on using the `IHostingEnvironment` service into it's own section ("Determining the environment at runtime") prior to discussing the Startup conventions. `IHostingEnvironment` provides the core abstraction for working with environments, so I think it deserves more direct attention - the startup conventions are really just sugar on top. We should specifically say that `IHostingEnvironment` is a *service* provided by the ASP.NET hosting layer that can be *injected* into your startup logic.